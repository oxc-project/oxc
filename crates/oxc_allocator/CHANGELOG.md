# Changelog

All notable changes to this crate will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project does not adhere to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) until v1.0.0.

## [0.6.0] - 2024-02-03

### Documentation

- Document behaviour of `Box`

## [0.5.0] - 2024-01-12

### Features

- Add `From` API (#1908)

## [0.4.0] - 2023-12-08

### Refactor

- Move to workspace lint table (#1444)

## [0.2.0] - 2023-09-14

### Refactor

- Improve code coverage a little bit
- Improve code coverage in various places (#721)

