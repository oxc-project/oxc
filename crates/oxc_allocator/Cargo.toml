[package]
name                   = "oxc_allocator"
version                = "0.11.0"
authors.workspace      = true
description.workspace  = true
edition.workspace      = true
homepage.workspace     = true
keywords.workspace     = true
license.workspace      = true
repository.workspace   = true
rust-version.workspace = true
categories.workspace   = true

[lints]
workspace = true

[lib]
doctest = false

[dependencies]
bumpalo = { workspace = true, features = ["collections"] }
serde   = { workspace = true }

[dev-dependencies]
serde_json = { workspace = true }
