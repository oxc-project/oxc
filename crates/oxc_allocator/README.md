# Oxc Allocator

A high-performance memory allocator using bump-based arena allocation for fast AST node creation.

## Overview

Oxc uses a bump-based memory arena for faster AST allocations. This crate provides an [`Allocator`] that manages memory efficiently by allocating objects in a single contiguous memory region, avoiding the overhead of individual heap allocations.

## Key Features

- **Bump allocation**: Fast allocation by simply incrementing a pointer
- **Arena-based memory management**: All allocations live for the same lifetime
- **Zero-copy data structures**: Efficient `Box`, `Vec`, `String`, and `HashMap` implementations
- **Optimal for AST operations**: Perfect for parse-transform-emit workflows

## Architecture

The allocator is designed specifically for AST processing workflows where:

1. A large number of nodes are allocated during parsing
2. All nodes have the same lifetime (tied to the AST)
3. Memory is released all at once when the AST is dropped

This approach is significantly faster than using the system allocator for AST operations.

## Features

- `serialize` - Enables serialization support for `Box` and `Vec` with `serde`
- `from_raw_parts` - Adds unsafe `from_raw_parts` method (not recommended for general use)
