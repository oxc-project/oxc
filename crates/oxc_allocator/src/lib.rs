use std::{
    convert::From,
    ops::{Deref, DerefMut},
};

mod arena;

pub use arena::{Box, String, Vec};
use bumpalo::Bump;

#[derive(Default)]
pub struct Allocator {
    bump: Bump,
}

impl From<Bump> for Allocator {
    fn from(bump: Bump) -> Self {
        Self { bump }
    }
}

impl Deref for Allocator {
    type Target = Bump;

    fn deref(&self) -> &Self::Target {
        &self.bump
    }
}

impl DerefMut for Allocator {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.bump
    }
}

#[cfg(test)]
mod test {
    use std::ops::Deref;

    use crate::Allocator;
    use bumpalo::Bump;

    #[test]
    fn test_api() {
        let bump = Bump::new();
        let allocator: Allocator = bump.into();
        #[allow(clippy::explicit_deref_methods)]
        {
            _ = allocator.deref();
        }
    }
}
