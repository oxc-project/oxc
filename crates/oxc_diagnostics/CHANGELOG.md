# Changelog

All notable changes to this crate will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project does not adhere to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) until v1.0.0.

## [0.12.3] - 2024-04-11

### Bug Fixes

- If format is json do not print summary information (#2899) (#2925)

## [0.10.0] - 2024-03-14

### Features

- Miette v7 (#2465)

## [0.8.0] - 2024-02-26

### Features

- Implement json reporter (#2452)
- Add json reporter (#2451)

## [0.6.0] - 2024-02-03

### Refactor

- Move all miette usages to `oxc_diagnostics`

## [0.5.0] - 2024-01-12

### Bug Fixes

- Always print without considering the `--max-warnings` option (#1996)

## [0.4.0] - 2023-12-08

### Refactor

- Move to workspace lint table (#1444)

## [0.2.0] - 2023-09-14

### Bug Fixes

- Use one version of `textwrap`
- Fix race condition when resolving paths

### Refactor

- Move path processing logic from cli to linter (#766)
- Implement DiagnosticService (#762)

