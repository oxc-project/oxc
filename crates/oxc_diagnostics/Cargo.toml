[package]
name                   = "oxc_diagnostics"
version                = "0.4.0"
authors.workspace      = true
description.workspace  = true
edition.workspace      = true
homepage.workspace     = true
keywords.workspace     = true
license.workspace      = true
repository.workspace   = true
rust-version.workspace = true
categories.workspace   = true

[lints]
workspace = true

[lib]
doctest = false

[dependencies]
thiserror = { workspace = true }
miette    = { workspace = true }

unicode-width = "0.1.11"
owo-colors    = { version = "3.5.0" }
textwrap      = { version = "=0.15.0" } # pinned to 0.15.0 to sync with miette
is-terminal   = { version = "0.4.10" }
