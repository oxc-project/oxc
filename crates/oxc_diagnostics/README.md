# Oxc Diagnostics

Error reporting and diagnostic utilities for JavaScript and TypeScript tooling.

## Overview

This crate provides comprehensive error handling and diagnostic reporting capabilities. It implements the [miette] diagnostic trait, making it compatible with other Rust diagnostic tooling while providing specialized features for JavaScript/TypeScript errors.

## Key Features

- **Rich diagnostics**: Detailed error messages with source context
- **Source highlighting**: Syntax-highlighted error locations
- **Multiple error support**: Collect and report multiple errors at once
- **Miette integration**: Compatible with the miette diagnostic ecosystem
- **Severity levels**: Support for errors, warnings, and informational messages

## Architecture

### Diagnostic Components

- **Message**: Primary error/warning description
- **Labels**: Highlight specific source locations
- **Help text**: Suggestions for fixing the issue
- **Source context**: Display relevant source code sections

### Error Flow

1. **Creation**: Tools create `OxcDiagnostic` instances for problems
2. **Collection**: `DiagnosticService` aggregates diagnostics
3. **Formatting**: Rich terminal output with colors and context
4. **Reporting**: Display to users in IDE, CLI, or other interfaces

The diagnostic system ensures consistent, high-quality error reporting across all oxc tools.
