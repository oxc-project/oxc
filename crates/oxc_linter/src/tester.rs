use std::{
    env,
    path::{Path, PathBuf},
};

use oxc_allocator::Allocator;
use oxc_diagnostics::miette::NamedSource;
use oxc_diagnostics::{DiagnosticService, GraphicalReportHandler, GraphicalTheme};
use serde::Deserialize;
use serde_json::Value;

use crate::{
    rules::RULES, ESLintSettings, Fixer, LintOptions, LintService, LintServiceOptions, Linter,
    RuleEnum,
};

#[derive(Eq, PartialEq)]
enum TestResult {
    Passed,
    Failed,
    Fixed(String),
}

#[derive(Debug, Clone, Default)]
pub struct TestCase {
    source: String,
    config: Option<Value>,
    settings: Option<Value>,
    path: Option<PathBuf>,
}

impl From<&str> for TestCase {
    fn from(source: &str) -> Self {
        Self { source: source.to_string(), ..Self::default() }
    }
}

impl From<(&str, Option<Value>)> for TestCase {
    fn from((source, config): (&str, Option<Value>)) -> Self {
        Self { source: source.to_string(), config, ..Self::default() }
    }
}

impl From<(&str, Option<Value>, Option<Value>)> for TestCase {
    fn from((source, config, settings): (&str, Option<Value>, Option<Value>)) -> Self {
        Self { source: source.to_string(), config, settings, ..Self::default() }
    }
}

impl From<(&str, Option<Value>, Option<Value>, Option<PathBuf>)> for TestCase {
    fn from(
        (source, config, settings, path): (&str, Option<Value>, Option<Value>, Option<PathBuf>),
    ) -> Self {
        Self { source: source.to_string(), config, settings, path }
    }
}

pub struct Tester {
    rule_name: &'static str,
    rule_path: PathBuf,
    expect_pass: Vec<TestCase>,
    expect_fail: Vec<TestCase>,
    expect_fix: Vec<(String, String, Option<Value>)>,
    snapshot: String,
    current_working_directory: Box<Path>,
    import_plugin: bool,
    jest_plugin: bool,
    jsx_a11y_plugin: bool,
    nextjs_plugin: bool,
    react_perf_plugin: bool,
}

impl Tester {
    pub fn new<T: Into<TestCase>>(
        rule_name: &'static str,
        expect_pass: Vec<T>,
        expect_fail: Vec<T>,
    ) -> Self {
        let rule_path = PathBuf::from(rule_name.replace('-', "_")).with_extension("tsx");
        let expect_pass = expect_pass.into_iter().map(Into::into).collect::<Vec<_>>();
        let expect_fail = expect_fail.into_iter().map(Into::into).collect::<Vec<_>>();
        let current_working_directory =
            env::current_dir().unwrap().join("fixtures/import").into_boxed_path();
        Self {
            rule_name,
            rule_path,
            expect_pass,
            expect_fail,
            expect_fix: vec![],
            snapshot: String::new(),
            current_working_directory,
            import_plugin: false,
            jest_plugin: false,
            jsx_a11y_plugin: false,
            nextjs_plugin: false,
            react_perf_plugin: false,
        }
    }

    /// Change the path
    pub fn change_rule_path(mut self, path: &str) -> Self {
        self.rule_path = self.current_working_directory.join(path);
        self
    }

    pub fn with_import_plugin(mut self, yes: bool) -> Self {
        self.import_plugin = yes;
        self
    }

    pub fn with_jest_plugin(mut self, yes: bool) -> Self {
        self.jest_plugin = yes;
        self
    }

    pub fn with_jsx_a11y_plugin(mut self, yes: bool) -> Self {
        self.jsx_a11y_plugin = yes;
        self
    }

    pub fn with_nextjs_plugin(mut self, yes: bool) -> Self {
        self.nextjs_plugin = yes;
        self
    }

    pub fn with_react_perf_plugin(mut self, yes: bool) -> Self {
        self.react_perf_plugin = yes;
        self
    }

    pub fn expect_fix<S: Into<String>>(mut self, expect_fix: Vec<(S, S, Option<Value>)>) -> Self {
        self.expect_fix =
            expect_fix.into_iter().map(|(s1, s2, r)| (s1.into(), s2.into(), r)).collect::<Vec<_>>();
        self
    }

    pub fn test(&mut self) {
        self.test_pass();
        self.test_fail();
        self.test_fix();
    }

    pub fn test_and_snapshot(&mut self) {
        self.test();
        self.snapshot();
    }

    pub fn snapshot(&self) {
        let name = self.rule_name.replace('-', "_");
        insta::with_settings!({ prepend_module_to_snapshot => false, }, {
            insta::assert_snapshot!(name.clone(), self.snapshot, &name);
        });
    }

    fn test_pass(&mut self) {
        for TestCase { source, config, settings, path } in self.expect_pass.clone() {
            let result = self.run(&source, config, false, &settings, &path);
            let passed = result == TestResult::Passed;
            assert!(passed, "expect test to pass: {source} {}", self.snapshot);
        }
    }

    fn test_fail(&mut self) {
        for TestCase { source, config, settings, path } in self.expect_fail.clone() {
            let result = self.run(&source, config, false, &settings, &path);
            let failed = result == TestResult::Failed;
            assert!(failed, "expect test to fail: {source}");
        }
    }

    fn test_fix(&mut self) {
        for (test, expected, config) in self.expect_fix.clone() {
            let result = self.run(&test, config, true, &None, &None);
            if let TestResult::Fixed(fixed_str) = result {
                assert_eq!(expected, fixed_str);
            } else {
                unreachable!()
            }
        }
    }

    fn run(
        &mut self,
        source_text: &str,
        config: Option<Value>,
        is_fix: bool,
        settings: &Option<Value>,
        path: &Option<PathBuf>,
    ) -> TestResult {
        let allocator = Allocator::default();
        let rule = self.find_rule().read_json(config);
        let lint_settings: ESLintSettings = settings
            .as_ref()
            .map_or_else(ESLintSettings::default, |v| ESLintSettings::deserialize(v).unwrap());
        let options = LintOptions::default()
            .with_fix(is_fix)
            .with_import_plugin(self.import_plugin)
            .with_jest_plugin(self.jest_plugin)
            .with_jsx_a11y_plugin(self.jsx_a11y_plugin)
            .with_nextjs_plugin(self.nextjs_plugin)
            .with_react_perf_plugin(self.react_perf_plugin);
        let linter = Linter::from_options(options)
            .unwrap()
            .with_rules(vec![rule])
            .with_settings(lint_settings);
        let path_to_lint = if self.import_plugin {
            assert!(path.is_none(), "import plugin does not support path");
            self.current_working_directory.join(&self.rule_path)
        } else if let Some(path) = path {
            self.current_working_directory.join(path)
        } else {
            self.rule_path.clone()
        };

        let cwd = self.current_working_directory.clone();
        let paths = vec![path_to_lint.into_boxed_path()];
        let options = LintServiceOptions { cwd, paths, tsconfig: None };
        let lint_service = LintService::from_linter(linter, options);
        let diagnostic_service = DiagnosticService::default();
        let tx_error = diagnostic_service.sender();
        let result = lint_service.run_source(&allocator, source_text, false, tx_error);

        if result.is_empty() {
            return TestResult::Passed;
        }

        if is_fix {
            let fix_result = Fixer::new(source_text, result).fix();
            return TestResult::Fixed(fix_result.fixed_code.to_string());
        }

        let diagnostic_path = if self.import_plugin {
            self.rule_path.strip_prefix(&self.current_working_directory).unwrap()
        } else {
            &self.rule_path
        }
        .to_string_lossy();

        let handler = GraphicalReportHandler::new().with_theme(GraphicalTheme::unicode_nocolor());
        for diagnostic in result {
            let diagnostic = diagnostic.error.with_source_code(NamedSource::new(
                diagnostic_path.clone(),
                source_text.to_string(),
            ));
            handler.render_report(&mut self.snapshot, diagnostic.as_ref()).unwrap();
        }
        TestResult::Failed
    }

    fn find_rule(&self) -> &RuleEnum {
        RULES
            .iter()
            .find(|rule| rule.name() == self.rule_name)
            .unwrap_or_else(|| panic!("Rule not found: {}", &self.rule_name))
    }
}
