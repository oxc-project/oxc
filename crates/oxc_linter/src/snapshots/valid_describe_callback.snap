---
source: crates/oxc_linter/src/tester.rs
---
  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:1]
 1 │ describe.each()()
   · ─────────────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:1]
 1 │ describe['each']()()
   · ────────────────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:1]
 1 │ describe.each(() => {})()
   · ─────────────────────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:25]
 1 │ describe.each(() => {})('foo')
   ·                         ─────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe.each()(() => {})
   ·                 ────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:20]
 1 │ describe['each']()(() => {})
   ·                    ────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:22]
 1 │ describe.each('foo')(() => {})
   ·                      ────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:27]
 1 │ describe.only.each('foo')(() => {})
   ·                           ────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:10]
 1 │ describe(() => {})
   ·          ────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:10]
 1 │ describe('foo')
   ·          ─────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Second argument must be a function
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe('foo', 'foo2')
   ·                 ──────
   ╰────
  help: Replace second argument with a function

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:1]
 1 │ describe()
   · ──────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe('foo', async () => {})
   ·                 ──────────────
   ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe('foo', async function () {})
   ·                 ────────────────────
   ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
   ╭─[valid_describe_callback.tsx:1:18]
 1 │ xdescribe('foo', async function () {})
   ·                  ────────────────────
   ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
   ╭─[valid_describe_callback.tsx:1:18]
 1 │ fdescribe('foo', async function () {})
   ·                  ────────────────────
   ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
   ╭─[valid_describe_callback.tsx:3:30]
 2 │             import { fdescribe } from '@jest/globals';
 3 │             fdescribe('foo', async function () {})
   ·                              ────────────────────
 4 │             
   ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
   ╭─[valid_describe_callback.tsx:1:22]
 1 │ describe.only('foo', async function () {})
   ·                      ────────────────────
   ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
   ╭─[valid_describe_callback.tsx:1:22]
 1 │ describe.skip('foo', async function () {})
   ·                      ────────────────────
   ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
    ╭─[valid_describe_callback.tsx:6:35]
  5 │                     });
  6 │ ╭─▶                 describe('async', async () => {
  7 │ │                       await new Promise(setImmediate);
  8 │ │                       it('breaks', () => {
  9 │ │                           throw new Error('Fail');
 10 │ │                       });
 11 │ ╰─▶                 });
 12 │                 });
    ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected return statement in describe callback
   ╭─[valid_describe_callback.tsx:3:17]
 2 │                 describe('foo', function () {
 3 │ ╭─▶                 return Promise.resolve().then(() => {
 4 │ │                       it('breaks', () => {
 5 │ │                           throw new Error('Fail')
 6 │ │                       })
 7 │ ╰─▶                 })
 8 │                 })
   ╰────
  help: Remove return statement in your describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected return statement in describe callback
    ╭─[valid_describe_callback.tsx:9:21]
  8 │                     describe('nested', () => {
  9 │ ╭─▶                     return Promise.resolve().then(() => {
 10 │ │                           it('breaks', () => {
 11 │ │                               throw new Error('Fail')
 12 │ │                           })
 13 │ ╰─▶                     })
 14 │                     })
    ╰────
  help: Remove return statement in your describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected return statement in describe callback
   ╭─[valid_describe_callback.tsx:3:17]
 2 │                 describe('foo', () => {
 3 │ ╭─▶                 return Promise.resolve().then(() => {
 4 │ │                       it('breaks', () => {
 5 │ │                           throw new Error('Fail')
 6 │ │                       })
 7 │ ╰─▶                 })
 8 │                     describe('nested', () => {
   ╰────
  help: Remove return statement in your describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected return statement in describe callback
    ╭─[valid_describe_callback.tsx:6:21]
  5 │                     describe('nested', () => {
  6 │ ╭─▶                     return Promise.resolve().then(() => {
  7 │ │                           it('breaks', () => {
  8 │ │                               throw new Error('Fail')
  9 │ │                           })
 10 │ ╰─▶                     })
 11 │                     })
    ╰────
  help: Remove return statement in your describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
    ╭─[valid_describe_callback.tsx:2:29]
  1 │     
  2 │ ╭─▶             describe('foo', async () => {
  3 │ │                   await something()
  4 │ │                   it('does something')
  5 │ │                   describe('nested', () => {
  6 │ │                       return Promise.resolve().then(() => {
  7 │ │                           it('breaks', () => {
  8 │ │                               throw new Error('Fail')
  9 │ │                           })
 10 │ │                       })
 11 │ │                   })
 12 │ ╰─▶             })
 13 │                 
    ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected return statement in describe callback
   ╭─[valid_describe_callback.tsx:1:23]
 1 │ describe('foo', () => test('bar', () => {})) 
   ·                       ─────────────────────
   ╰────
  help: Remove return statement in your describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected argument(s) in describe callback
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe('foo', done => {})
   ·                 ──────────
   ╰────
  help: Remove argument(s) of describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected argument(s) in describe callback
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe('foo', function (done) {})
   ·                 ──────────────────
   ╰────
  help: Remove argument(s) of describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected argument(s) in describe callback
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe('foo', function (one, two, three) {})
   ·                 ─────────────────────────────
   ╰────
  help: Remove argument(s) of describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe('foo', async function (done) {})
   ·                 ────────────────────────
   ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected argument(s) in describe callback
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe('foo', async function (done) {})
   ·                 ────────────────────────
   ╰────
  help: Remove argument(s) of describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:1]
 1 │ describe.each()()
   · ─────────────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:1]
 1 │ describe["each"]()()
   · ────────────────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:1]
 1 │ describe.each(() => {})()
   · ─────────────────────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:25]
 1 │ describe.each(() => {})("foo")
   ·                         ─────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe.each()(() => {})
   ·                 ────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:20]
 1 │ describe["each"]()(() => {})
   ·                    ────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:22]
 1 │ describe.each("foo")(() => {})
   ·                      ────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:27]
 1 │ describe.only.each("foo")(() => {})
   ·                           ────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:10]
 1 │ describe(() => {})
   ·          ────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:10]
 1 │ describe("foo")
   ·          ─────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): Second argument must be a function
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe("foo", "foo2")
   ·                 ──────
   ╰────
  help: Replace second argument with a function

  ⚠ eslint-plugin-vitest(valid-describe-callback): Describe requires name and callback arguments
   ╭─[valid_describe_callback.tsx:1:1]
 1 │ describe()
   · ──────────
   ╰────
  help: Add name as first argument and callback as second argument

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe("foo", async () => {})
   ·                 ──────────────
   ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe("foo", async function () {})
   ·                 ────────────────────
   ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
   ╭─[valid_describe_callback.tsx:1:18]
 1 │ xdescribe("foo", async function () {})
   ·                  ────────────────────
   ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
   ╭─[valid_describe_callback.tsx:1:18]
 1 │ fdescribe("foo", async function () {})
   ·                  ────────────────────
   ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
   ╭─[valid_describe_callback.tsx:1:22]
 1 │ describe.only("foo", async function () {})
   ·                      ────────────────────
   ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
   ╭─[valid_describe_callback.tsx:1:22]
 1 │ describe.skip("foo", async function () {})
   ·                      ────────────────────
   ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
    ╭─[valid_describe_callback.tsx:6:39]
  5 │                         });
  6 │ ╭─▶                     describe('async', async () => {
  7 │ │                           await new Promise(setImmediate);
  8 │ │                           it('breaks', () => {
  9 │ │                               throw new Error('Fail');
 10 │ │                           });
 11 │ ╰─▶                     });
 12 │                     });
    ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected return statement in describe callback
   ╭─[valid_describe_callback.tsx:3:21]
 2 │                     describe('foo', function () {
 3 │ ╭─▶                     return Promise.resolve().then(() => {
 4 │ │                           it('breaks', () => {
 5 │ │                               throw new Error('Fail')
 6 │ │                           })
 7 │ ╰─▶                     })
 8 │                     })
   ╰────
  help: Remove return statement in your describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected return statement in describe callback
    ╭─[valid_describe_callback.tsx:9:25]
  8 │                         describe('nested', () => {
  9 │ ╭─▶                         return Promise.resolve().then(() => {
 10 │ │                               it('breaks', () => {
 11 │ │                                   throw new Error('Fail')
 12 │ │                               })
 13 │ ╰─▶                         })
 14 │                         })
    ╰────
  help: Remove return statement in your describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected return statement in describe callback
   ╭─[valid_describe_callback.tsx:3:21]
 2 │                     describe('foo', () => {
 3 │ ╭─▶                     return Promise.resolve().then(() => {
 4 │ │                           it('breaks', () => {
 5 │ │                               throw new Error('Fail')
 6 │ │                           })
 7 │ ╰─▶                     })
 8 │                         describe('nested', () => {
   ╰────
  help: Remove return statement in your describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected return statement in describe callback
    ╭─[valid_describe_callback.tsx:6:25]
  5 │                         describe('nested', () => {
  6 │ ╭─▶                         return Promise.resolve().then(() => {
  7 │ │                               it('breaks', () => {
  8 │ │                                   throw new Error('Fail')
  9 │ │                               })
 10 │ ╰─▶                         })
 11 │                         })
    ╰────
  help: Remove return statement in your describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
    ╭─[valid_describe_callback.tsx:2:33]
  1 │     
  2 │ ╭─▶                 describe('foo', async () => {
  3 │ │                       await something()
  4 │ │                       it('does something')
  5 │ │                       describe('nested', () => {
  6 │ │                           return Promise.resolve().then(() => {
  7 │ │                               it('breaks', () => {
  8 │ │                                   throw new Error('Fail')
  9 │ │                               })
 10 │ │                           })
 11 │ │                       })
 12 │ ╰─▶                 })
 13 │                 
    ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected return statement in describe callback
   ╭─[valid_describe_callback.tsx:3:21]
 2 │                 describe('foo', () =>
 3 │                     test('bar', () => {})
   ·                     ─────────────────────
 4 │                 )
   ╰────
  help: Remove return statement in your describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected argument(s) in describe callback
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe("foo", done => {})
   ·                 ──────────
   ╰────
  help: Remove argument(s) of describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected argument(s) in describe callback
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe("foo", function (done) {})
   ·                 ──────────────────
   ╰────
  help: Remove argument(s) of describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected argument(s) in describe callback
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe("foo", function (one, two, three) {})
   ·                 ─────────────────────────────
   ╰────
  help: Remove argument(s) of describe callback

  ⚠ eslint-plugin-vitest(valid-describe-callback): No async describe callback
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe("foo", async function (done) {})
   ·                 ────────────────────────
   ╰────
  help: Remove `async` keyword

  ⚠ eslint-plugin-vitest(valid-describe-callback): Unexpected argument(s) in describe callback
   ╭─[valid_describe_callback.tsx:1:17]
 1 │ describe("foo", async function (done) {})
   ·                 ────────────────────────
   ╰────
  help: Remove argument(s) of describe callback
